driver/tab_completion.py

```diff
@@ -0,0 +1,122 @@
+    if not builds_folder.exists():
+        argcomplete.warn("No build exists.")
+        exit(1)
+    return [p.name for p in builds_folder.iterdir() if p.is_dir()]
+
+
+def complete_planner_args(prefix, parsed_args, **kwargs):
+    if parsed_args.build and parsed_args.debug:
+        argcomplete.warn("The option --debug is an alias for --build=debug. Do no specify both --debug and --build.")
+        exit(1)
+
+    build = parsed_args.build
+    if not build:
+        if parsed_args.debug:
+            build = "debug"
+        else:
+            build = "release"
+
+    # Get some information from planner_args before it is deleted in split_planner_args().
+    planner_args = parsed_args.planner_args
+    num_planner_args = len(planner_args)
+    mode_switches = ["--translate-options", "--search-options"]
+    last_option_was_mode_switch = planner_args and (planner_args[-1] in mode_switches)
+    double_dash_in_options = "--" in planner_args
+
+    current_mode = util.split_planner_args(parsed_args)
+    num_filenames = len(parsed_args.filenames)
+    has_only_filename_options = (num_filenames == num_planner_args)
+    has_only_filename_or_double_dash_options = (num_filenames + int(double_dash_in_options) == num_planner_args)
+    can_use_double_dash = (1 <= num_planner_args <= 2) and has_only_filename_or_double_dash_options
+
+    completions = []
+
+    if can_use_double_dash:
+        completions.append("--")
+
+    if parsed_args.filenames or double_dash_in_options:
+        bin_dir = Path(util.REPO_ROOT_DIR) / "builds" / build / "bin"
+        if current_mode == "search":
+            if not last_option_was_mode_switch:
+                completions.append("--translate-options")
+
+            downward = bin_dir / "downward"
+            if downward.exists():
+                completions += get_completions_from_downward(
+                    downward, parsed_args.search_options, prefix)
+        else:
+            if not last_option_was_mode_switch:
+                completions.append("--search-options")
+
+            translator = bin_dir / "translate" / "translate.py"
+            if translator.exists():
+                completions += get_completions_from_translator(
+                    translator, parsed_args.translate_options, prefix)
+
+    if has_only_filename_options and len(parsed_args.filenames) < 2:
+        file_completer = argcomplete.FilesCompleter()
+        completions += file_completer(prefix, **kwargs)
+
+    return completions
+
+
+def get_completions_from_downward(downward, options, prefix):
+    simulated_commandline = [str(downward)] + options + [prefix]
+    comp_line = " ".join(simulated_commandline)
+    comp_point = str(len(comp_line))
+    comp_cword = str(len(simulated_commandline) - 1)
+    cmd = [str(downward), "--bash-complete",
+           comp_point, comp_line, comp_cword] + simulated_commandline
+    output = subprocess.check_output(cmd, text=True)
+    return output.split()
```

Comment:
Will the completions never contain spaces? For example if the user included a space in the argument to be completed?

In shell scripts, one would normally try to deal with this by using newlines for splitting (you often see IFS set to newline for this reason), which is also potentially problematic because arguments can use newlines, but is an improvement over splitting on all whitespace.

Reply:
The surrounding code also uses newlines (\n) as separators. We changed from `split` to `splitlines`. So far, the search component does not return any newlines in the suggestions. Once we implement the suggestions for `--search`, it could be part of the prefix and we will deal with it.